### PR TITLE
Support toplevel interpolations in graphql tags

### LINKED ISSUE #3318: Support top-level interpolations in gql tags
apollo supports _toplevel_ interpolations in `` gql`...` `` tagged template literals (it throws an error if you try to use interpolations in nested situations). This allows sharing GraphQL fragments between queries without loosing static analysis (both eslint-plugin-graphql and apollo-codegen support this).

Currently, we skip gql tags if they contain interpolations. It would be awesome if we could support toplevel ones though!

My idea is that we should _try_ to parse and print every literal part independently. If _all_ succeeds, we join the results up with the interpolations. If any of them fails, we bail out like today.

What do you think?

**Prettier 1.8.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBFArnAJwE9MBeTAcwEcAbAAwB0or9jMBVNQzYJzTXFwIAKAJYATJAxAAjCDOkBKXlH78AdJs6EAInBgBDUTTR81m9QDECohONOrMAXyYuoTACTB2AZQCiAEoA+jp+ACoAggCSADI+QZYBEQDiALJ+AHJhTl6JUZk68Ykp6VludCAANCAQAA4woujIoAYYyABmBiZw1TIEBmAA1vo+tQOiUBTIMAT41eIQYB1dXNUTQjAACv0UALYGy93VAFZoAB4AQv1DIwa7cDETcIerNbgwte8ATC89IGMEITIWQGGREGjQKr-GywADqEhgAAtkAAOAAM1VqBAgXFh-Vq01mfwIcBYohJ2wMewOSE6RxAXF2okJcwZEwoNDgeAg8F+1UMMnh4iRyC+-P6xnZAGEILt9sCoNBntVBHAwqC0L8nE4gA)

**Input:**
```jsx
const Query = gql`
query User {
  user(id:"bob"){
    ...UserDetails
    ...Friends
  }
}

${USER_DETAILS_FRAGMENT}${FRIENDS_FRAGMENT}
`
```

**Output:**
```jsx
const Query = gql`
query User {
  user(id:"bob"){
    ...UserDetails
    ...Friends
  }
}

${USER_DETAILS_FRAGMENT}${FRIENDS_FRAGMENT}
`;

```

**Expected behavior:**

```jsx
const Query = gql`
  query User {
    user(id: "bob") {
      ...UserDetails
      ...Friends
    }
  }

  ${USER_DETAILS_FRAGMENT}
  ${FRIENDS_FRAGMENT}
`;
```

### PR DESCRIPTION
**Pull request by lydell**
Fixes #3318.

### CODE REVIEW

File: tests/multiparser_js_graphql/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -82,6 +78,153 @@ const query = gql\`
...
+    test # comment
+  } # comment
+  \${test} # comment
+  \${test} # comment
+
+  \${test} # comment
+
+  # comment
+  \${test} # comment
+\`
```
Comment (lydell) [inline tests/multiparser_js_graphql/__snapshots__/jsfmt.spec.js.snap]: All the comments after the interpolations here are currently moved to the next line. Is this a deal-breaker? This seems to need some larger changes to change.

Comment (azz) [reply tests/multiparser_js_graphql/__snapshots__/jsfmt.spec.js.snap]: Probably not a deal-breaker. I'm guessing comments are pretty uncommon here. I've only ever used comments in GraphQL schemas.

**Comment by azz** [comment]
Multiparser refactor looks good. Haven't reviewed GraphQL changes properly yet but test outputs look good.

**Comment by lydell** [comment]
I need to figure out why the `parentParser` thing doesn’t work, though.

**Comment by lydell** [comment]
I need to figure out why the `parentParser` thing doesn’t work, though.

File: src/multiparser.js [hunk 2]
```diff
@@ -5,44 +5,38 @@ const docUtils = require("./doc-utils");
...
-      return fromBabylonFlowOrTypeScript(path, options);
+      return fromBabylonFlowOrTypeScript(path, print, options);
     case "markdown":
-      return fromMarkdown(path, options);
+      return fromMarkdown(path, print, options);
   }
 }
 
-function fromMarkdown(path, options) {
+function parseAndPrint(passedOptions) {
```
Comment (duailibe) [inline src/multiparser.js]: You'd need to pass the previous and next options here so the `parentParser` thing works.. see `printSubtree` how it builds `next.options` but gets the `parentParser` from `options`

**Comment by lydell** [comment]
I think I'm happy with this now. Would be awesome if it could make it into the 1.9 release!

**Comment by lydell** [comment]
> Haven't reviewed GraphQL changes properly yet 

@azz Do you want to take a look before merging, or should I just press the green button?

**Pull request merged**